### PR TITLE
Support for kubernetes clusters with user assigned msi 

### PR DESCRIPTION
**Pull request by kkmsft**
**What this PR does / why we need it**:

Support for kubernetes clusters with user assigned msi 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

- Support for kubernetes clusters with user assigned msi 
- Adds the go sdk files for msi apis.
- Avoid role assignment retry if the its already present.

**Special notes for your reviewer**:

**If applicable**:
- [ ] documentation
- [ ] unit tests
- [ ] tested backward compatibility (ie. deploy with previous version, upgrade with this branch)

**Release note**:

```release-note
```

### CODE REVIEW

File: examples/kubernetes-msi-userassigned/kube-vma.json [hunk 1]
```diff
@@ -0,0 +1,37 @@
+{
+  "apiVersion": "vlabs",
+  "properties": {
+    "orchestratorProfile": {
+      "orchestratorType": "Kubernetes",
+      "orchestratorRelease": "1.10",
+      "kubernetesConfig": {
+        "useManagedIdentity": true,
+	"userAssignedID": "acsenginetestid"
```
Comment (jackfrancis) [inline examples/kubernetes-msi-userassigned/kube-vma.json]: nit: spaces

File: examples/kubernetes-msi-userassigned/kube-vmss.json [hunk 2]
```diff
@@ -0,0 +1,37 @@
+{
+  "apiVersion": "vlabs",
+  "properties": {
+    "orchestratorProfile": {
+      "orchestratorType": "Kubernetes",
+      "orchestratorRelease": "1.10",
+      "kubernetesConfig": {
+        "useManagedIdentity": true,
+	"userAssignedID": "acsenginetestid"
```
Comment (jackfrancis) [inline examples/kubernetes-msi-userassigned/kube-vmss.json]: ditto

File: parts/k8s/kubernetesagentresourcesvmas.t [hunk 3]
```diff
@@ -146,10 +146,14 @@
     },
 {{end}}
   {
+    {{if UserAssignedIDEnabled}}
+     "apiVersion": "[variables('apiVersionUserMSI')]",
```
Comment (jackfrancis) [inline parts/k8s/kubernetesagentresourcesvmas.t]: Does the API version above work w/ managed disks? The way this is implemented it assumes that the user MSI API version takes precedence over the managed disks API version.

File: parts/k8s/kubernetesagentresourcesvmas.t [hunk 4]
```diff
@@ -255,7 +268,8 @@
       },
       "type": "Microsoft.Compute/virtualMachines"
     },
-    {{if UseManagedIdentity}}
+    {{if UseManagedIdentity }}
```
Comment (jackfrancis) [inline parts/k8s/kubernetesagentresourcesvmas.t]: nit: remove this add'l space so the changeset is easier to parse

File: parts/k8s/kubernetesconfigs.sh [hunk 5]
```diff
@@ -149,6 +149,7 @@ function configureK8s() {
     "cloudProviderRateLimitQPS": ${CLOUDPROVIDER_RATELIMIT_QPS},
     "cloudProviderRateLimitBucket": ${CLOUDPROVIDER_RATELIMIT_BUCKET},
     "useManagedIdentityExtension": ${USE_MANAGED_IDENTITY_EXTENSION},
+    "userAssignedIdentityID": "${USE_ASSIGNED_IDENTITY_ID}",
```
Comment (jackfrancis) [inline parts/k8s/kubernetesconfigs.sh]: `s/{USE_ASSIGNED_IDENTITY_ID}/{USER_ASSIGNED_IDENTITY_ID}` (note the add'l "R" in `USER`)

File: parts/k8s/kubernetesmastervars.t [hunk 6]
```diff
@@ -134,7 +139,7 @@
...
...roviderConfig').cloudProviderBackoffRetries,' CLOUDPROVIDER_BACKOFF_EXPONENT=',parameters('cloudproviderConfig').cloudProviderBackoffExponent,' CLOUDPROVIDER_BACKOFF_DURATION=',parameters('cloudproviderConfig').cloudProviderBackoffDuration,' CLOUDPROVIDER_BACKOFF_JITTER=',parameters('cloudproviderConfig').cloudProviderBackoffJitter,' CLOUDPROVIDER_RATELIMIT=',parameters('cloudproviderConfig').cloudProviderRatelimit,' CLOUDPROVIDER_RATELIMIT_QPS=',parameters('cloudproviderConfig').cloudProviderRatelimitQPS,' CLOUDPROVIDER_RATELIMIT_BUCKET=',parameters('cloudproviderConfig').cloudProviderRatelimitBucket,' USE_MANAGED_IDENTITY_EXTENSION=',variables('useManagedIdentityExtension'),' USE_ASSIGNED_IDENTITY_ID=',variables('userAssignedClientID'),' USE_INSTANCE_METADATA=',variables('useInstanceMetadata'),' LOAD_BALANCER_SKU=',variables('loadBalancerSku'),' EXCLUDE_MASTER_FROM_STANDARD_LB=',variables('excludeMasterFromStandardLB'),' CONTAINER_RUNTIME=',parameters('containerRuntime'),' CONTAINERD_DOWNLOAD_URL_BASE=',parameters('containerdDownloadURLBase'),' POD_INFRA_CONTAINER_SPEC=',parameters('kubernetesPodInfraContainerSpec'),' KMS_PROVIDER_VAULT_NAME=',variables('clusterKeyVaultName'))]",
```
Comment (jackfrancis) [inline parts/k8s/kubernetesmastervars.t]: see above comment about changing to `USER_ASSIGNED_IDENTITY_ID` (not the additional "R")

**Comment by jackfrancis** [comment]
Lint errors will help nudge unit tests:

```
pkg/armhelpers/msi.go:14:1:warning: exported method AzureClient.IsUserAssignedIDPresent should have comment or be unexported (golint)
pkg/armhelpers/msi.go:33:1:warning: exported method AzureClient.CreateUserAssignedID should have comment or be unexported (golint)
```

**Comment by jackfrancis** [comment]
Please update `Gopkg.toml` to reflect the addition of these new dependencies.

**Comment by jackfrancis** [comment]
This is looking well on its way, thanks @kkmsft! See comments.

File: parts/k8s/kubernetesagentresourcesvmas.t (hunk 3 above)
Comment (kkmsft) [reply parts/k8s/kubernetesagentresourcesvmas.t]: Testing was done with this and the managed disk set as below and it created the cluster fine. 

 "agentPoolProfiles": [
      {
        "name": "agentpool1",
        "count": 3,
        "vmSize": "Standard_D2_v2",
        "availabilityProfile": "VirtualMachineScaleSets",
        "storageProfile": "ManagedDisks"
      }

Is there a way to use one api version only for  parts of the specification - like for example to specify the identity in this case ?

**Comment by kkmsft** [comment]
@jackfrancis - not able to get the Gopkg.toml updated since this just includes packages (services/preview/msi/mgmt/2015-08-31-preview/msi) from the azure-sdk-for-go project. Will catch you offline and figure out what to do about this..

**Comment by jackfrancis** [comment]
Thanks @kkmsft!

**Pull request merged**